### PR TITLE
Activate example CI on Python3.9

### PR DESCRIPTION
**Pull request by crcrpar**
This pull request only activates PyTorch example, i.e.
`pytorch_simple.py`.
The change to `setup.py` is quite tentative and far from what this
should look like, however, as of now this change is enough given the
number of requirements: it's cumbersome to add the constraints all the
other packages in 'example'.

Related: #2034

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Could you review this PR, please?

File: setup.py [hunk 1]
```diff
@@ -112,6 +112,14 @@ def get_extras_require() -> Dict[str, List[str]]:
...
             "optax",
             "dm-haiku",
+        ]
+        if sys.version_info[:2] < (3, 9)
+        else [
+            "torch==1.7.1 ; sys_platform=='darwin'",
+            "torch==1.7.1+cpu ; sys_platform!='darwin'",
+            "torchvision==0.8.2 ; sys_platform=='darwin'",
+            "torchvision==0.8.2+cpu ; sys_platform!='darwin'",
+            "torchaudio==0.7.2",
```
Comment (toshihikoyanase) [inline setup.py:122]: If we manage the dependencies for Python 3.9 separately, the list of software will be doubled in the future. How about adding the version constraints as follows?
```python
        "example": [
            "catboost ; python_version<'3.9'",
            "chainer ; python_version<'3.9'",
            "lightgbm ; python_version<'3.9'",
            "mlflow ; python_version<'3.9'",
            "mpi4py ; python_version<'3.9'",
            "mxnet ; python_version<'3.9'",
            "nbval ; python_version<'3.9'",
            "scikit-image ; python_version<'3.9'",
            "scikit-learn>=0.19.0,<0.23.0 ; python_version<'3.9'",  # optuna/visualization/param_importances.py.
            "xgboost ; python_version<'3.9'",
            "keras ; python_version<'3.9'",
            "tensorflow>=2.0.0 ; python_version<'3.9'",
            "tensorflow-datasets ; python_version<'3.9'",
            "pytorch-ignite ; python_version<'3.9'",
            "pytorch-lightning>=1.0.2 ; python_version<'3.9'",
            "thop ; python_version<'3.9'",
            "skorch ; python_version<'3.9'",
            "stable-baselines3>=0.7.0 ; python_version<'3.9'",
            "catalyst ; python_version<'3.9'",
            "torch==1.7.1 ; sys_platform=='darwin'",
            "torch==1.7.1+cpu ; sys_platform!='darwin'",
            "torchvision==0.8.2 ; sys_platform=='darwin'",
            "torchvision==0.8.2+cpu ; sys_platform!='darwin'",
            "torchaudio==0.7.2",
            "allennlp==1.2.0 ; python_version<'3.9'",
            "dask[dataframe] ; python_version<'3.9'",
            "dask-ml ; python_version<'3.9'",
            "botorch ; python_version>'3.6' and python_version<'3.9'",
            "fastai ; python_version<'3.9'",
            "jax ; python_version<'3.9'",
            "optax ; python_version<'3.9'",
            "dm-haiku ; python_version<'3.9'",
        ],
```

File: .github/workflows/examples.yml [hunk 2]
```diff
@@ -54,6 +54,8 @@ jobs:
           IGNORES='chainermn_.*|fastai*|rapids_.*|botorch_.*'
         elif [ ${{ matrix.python-version }} = 3.8 ]; then
           IGNORES='chainermn_.*|dask_ml_.*|keras_.*|tensorboard_.*|tensorflow_.*|tfkeras_.*|fastai*|allennlp|rapids_.*'
+        elif [ ${{ matrix.python-version }} = 3.9 ]; then
+          IGNORES='botorch_*|catalyst_*|catboost_*|chainer_*|chainermn_.*|dask_ml_.*|fastai*|gluon_*|haiku_*|keras_*|lightgbm_*|mxnet_*|optuna_search_cv_*|pytorch_ignite_*|pytorch_lightning_*|rapids_*|skimage_lbp_*|sklearn_*|skorch_*|tensorboard_*|tensorflow_*|tfkeras_*|xgboost_*'
```
Comment (toshihikoyanase) [inline .github/workflows/examples.yml:58]: The stable-baseline example and the allennlp examples are not excluded from the target. Could you add them to the `IGNORES`?
```suggestion
          IGNORES='allennlp_*|botorch_*|catalyst_*|catboost_*|chainer_*|chainermn_.*|dask_ml_.*|fastai*|gluon_*|haiku_*|keras_*|lightgbm_*|mxnet_*|optuna_search_cv_*|pytorch_ignite_*|pytorch_lightning_*|rapids_*|sb3_*|skimage_lbp_*|sklearn_*|skorch_*|tensorboard_*|tensorflow_*|tfkeras_*|xgboost_*'
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. It basically looks good to me, but I left small comments.

File: setup.py (hunk 1 above)
Comment (crcrpar) [reply setup.py:122]: I don't think that list whose most items come with the identical constraint among them.

What in my mind is to merge the separate two requirement lists after the Python3.9 list gets the bunch of items.

Comment (toshihikoyanase) [reply setup.py:122]: > What in my mind is to merge the separate two requirement lists after the Python3.9 list gets the bunch of items.

I see. Let's merge the lists after we add more packages to the Python 3.9 list.

**Comment by crcrpar** [comment]
TODO for @crcrpar: Make IGNORES more comprehensive.

File: .github/workflows/examples.yml [hunk 3]
```diff
@@ -54,6 +54,8 @@ jobs:
           IGNORES='chainermn_.*|fastai*|rapids_.*|botorch_.*'
         elif [ ${{ matrix.python-version }} = 3.8 ]; then
           IGNORES='chainermn_.*|dask_ml_.*|keras_.*|tensorboard_.*|tensorflow_.*|tfkeras_.*|fastai*|allennlp|rapids_.*'
+        elif [ ${{ matrix.python-version }} = 3.9 ]; then
+          IGNORES='allennlp_*|botorch_*|catalyst_*|catboost_*|chainer_*|chainermn_.*|dask_ml_.*|fastai*|gluon_*|haiku_*|keras_*|lightgbm_*|mxnet_*|optuna_search_cv_*|pytorch_ignite_*|pytorch_lightning_*|rapids_*|sb3_*|skimage_lbp_*|sklearn_*|skorch_*|tensorboard_*|tensorflow_*|tfkeras_*|xgboost_*'
```
Comment (toshihikoyanase) [inline .github/workflows/examples.yml:58]: The target examples will be as follows, but many of them did not work as expected due to the lack of dependencies. Could you exclude them?
- examples/pruning/simple.py: did not work due to the lack of `scikit-learn`
- examples/enqueue_trial.py: did not work due to the lack of `scikit-learn`
- examples/quadratic_simple.py: OK 
- examples/visualization/plot_pareto_front.py: did not work due to the lack of `plotly`
- examples/visualization/plot_study.py: did not work due to the lack of `plotly`
- examples/pytorch_simple.py: OK
- examples/multi_objective/pytorch_simple.py: did not work due to the lack of `thop`

Comment (keisuke-umezawa) [reply .github/workflows/examples.yml:58]: lgtm other than this comment.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! Thank you for committing!

File: .github/workflows/examples.yml [hunk 4]
```diff
@@ -53,7 +53,10 @@ jobs:
         if [ ${{ matrix.python-version }} = 3.6 ]; then
           IGNORES='chainermn_.*|fastai*|rapids_.*|botorch_.*'
         elif [ ${{ matrix.python-version }} = 3.8 ]; then
+          IGNORES='chainermn_.*|dask_ml_.*|keras_.*|tensorboard_.*|tensorflow_.*|tfkeras_.*|fastai*|allennlp|rapids_.*'
```
Comment (toshihikoyanase) [inline .github/workflows/examples.yml:56]: I think this line is duplicated with the next line. In addition, `allennlp` is supported by Python 3.8 according to #2226.  Could you remove this line?
```suggestion
```

File: .github/workflows/examples.yml [hunk 5]
```diff
@@ -53,7 +53,10 @@ jobs:
...
           IGNORES='chainermn_.*|dask_ml_.*|keras_.*|tensorboard_.*|tensorflow_.*|tfkeras_.*|fastai*|rapids_.*'
+        elif [ ${{ matrix.python-version }} = 3.9 ]; then
+          IGNORES='allennlp_*|botorch_*|catalyst_*|catboost_*|chainer_*|chainermn_.*|dask_ml_.*|fastai*|gluon_*|haiku_*|keras_*|lightgbm_*|mxnet_*|optuna_search_cv_*|pytorch_ignite_*|pytorch_lightning_*|rapids_*|sb3_*|skimage_lbp_*|sklearn_*|skorch_*|tensorboard_*|tensorflow_*|tfkeras_*|xgboost_*|pruning/simple*|enqueue_trial*|visualization/plot_pareto_front*|visualization/plot_study*|multi_objective/pytorch_simple*'
```
Comment (toshihikoyanase) [inline .github/workflows/examples.yml:59]: Could you add `hydra_*|` to `IGNORES`? It was introduced by #2290.
```suggestion
          IGNORES='allennlp_*|botorch_*|catalyst_*|catboost_*|chainer_*|chainermn_.*|dask_ml_.*|fastai*|gluon_*|haiku_*|hydra_*|keras_*|lightgbm_*|mxnet_*|optuna_search_cv_*|pytorch_ignite_*|pytorch_lightning_*|rapids_*|sb3_*|skimage_lbp_*|sklearn_*|skorch_*|tensorboard_*|tensorflow_*|tfkeras_*|xgboost_*|pruning/simple*|enqueue_trial*|visualization/plot_pareto_front*|visualization/plot_study*|multi_objective/pytorch_simple*'
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update. I added two comments.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!

**Pull request merged**